<center>
</center>

# Working with a real world data-set using SQL and Python

Estaimted time needed: **30** minutes

## Objectives

After complting this lab you will be able to:

*   Understand the dataset for Chicago Public School level performance
*   Store the dataset in an Db2 database on IBM Cloud instance
*   Retrieve metadata about tables and columns and query data from mixed case columns
*   Solve example problems to practice your SQL skills including using built-in database functions


## Chicago Public Schools - Progress Report Cards (2011-2012)

The city of Chicago released a dataset showing all school level performance data used to create School Report Cards for the 2011-2012 school year. The dataset is available from the Chicago Data Portal: [https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

This dataset includes a large number of metrics. Start by familiarizing yourself with the types of metrics in the database: [https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?download=true](https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&download=true&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

**NOTE**:

Do not download the dataset directly from City of Chicago portal. Instead download a static copy which is a more database friendly version from this <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">link</a>.

Now review some of its contents.


### Store the dataset in a Table

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. To analyze the data using SQL, it first needs to be stored in the database.

While it is easier to read the dataset into a Pandas dataframe and then PERSIST it into the database as we saw in the previous lab, it results in mapping to default datatypes which may not be optimal for SQL querying. For example a long textual field may map to a CLOB instead of a VARCHAR.

Therefore, **it is highly recommended to manually load the table using the database console LOAD tool, as indicated in Week 2 Lab 1 Part II**. The only difference with that lab is that in Step 5 of the instructions you will need to click on create "(+) New Table" and specify the name of the table you want to create and then click "Next".

##### Now open the Db2 console, open the LOAD tool, Select / Drag the .CSV file for the CHICAGO PUBLIC SCHOOLS dataset and load the dataset into a new table called **SCHOOLS**.

<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01"><img src = "https://ibm.box.com/shared/static/uc4xjh1uxcc78ks1i18v668simioz4es.jpg"></a>


### Connect to the database

Let us now load the ipython-sql  extension and establish a connection with the database

The following modules are pre-installed in the Skills Network Labs environment. However if you run this notebook commands in a different Jupyter environment (e.g. Watson Studio or Ananconda) you may need to install these libraries by removing the `#` sign before `!pip` in the code cell below.


In [1]:
# These libraries are pre-installed in SN Labs. If running in another environment please uncomment lines below to install them:
!pip install --force-reinstall ibm_db==3.1.0 ibm_db_sa==0.3.3
# Ensure we don't load_ext with sqlalchemy>=1.4 (incompadible)
!pip uninstall sqlalchemy==1.4 -y && pip install sqlalchemy==1.3.24
!pip install ipython-sql"

  Using cached https://files.pythonhosted.org/packages/f5/34/adc2134c9567dd99254f20e6981a9006a5767dfed287eb94f273ec51e092/greenlet-1.1.2-cp37-cp37m-win_amd64.whl
  Using cached https://files.pythonhosted.org/packages/58/fd/f82049e9b21773188a72bd1bc53530b900392808f8b823e4c778e7c035f0/importlib_metadata-4.10.1-py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/52/c5/df7953fe6065185af5956265e3b16f13c2826c2b1ba23d43154f3af453bc/zipp-3.7.0-py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/05/e4/baf0031e39cf545f0c9edd5b1a2ea12609b7fcba2d58e118b11753d68cf0/typing_extensions-4.0.1-py3-none-any.whl
  Found existing installation: ibm-db 3.1.0
    Uninstalling ibm-db-3.1.0:
      Successfully uninstalled ibm-db-3.1.0
  Found existing installation: greenlet 1.1.2
    Uninstalling greenlet-1.1.2:
      Successfully uninstalled greenlet-1.1.2
  Found existing installation: zipp 3.7.0
    Uninstalling zipp-3.7.0:
      Successfully uninstalled zipp-3.7.0
  F

In [54]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [55]:
# Enter the connection string for your Db2 on Cloud database instance below
# %sql ibm_db_sa://my-username:my-password@my-hostname:my-port/my-db-name?security=SSL
%sql ibm_db_sa://xqk06680:B9aecRgnyF2754zQ@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;

### Query the database system catalog to retrieve table metadata

##### You can verify that the table creation was successful by retrieving the list of all tables in your schema and checking whether the SCHOOLS table was created


In [42]:
# type in your query to retrieve list of all tables in the database for your db2 schema (username)
%sql select TABSCHEMA,TABNAME, CREATE_TIME from SYSCAT.TABLES where tabschema='XQK06680'

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('XQK06680', 'EMPLOYEES', datetime.datetime(2022, 1, 12, 22, 41, 7, 636023)),
 ('XQK06680', 'JOB_HISTORY', datetime.datetime(2022, 1, 12, 22, 41, 7, 807618)),
 ('XQK06680', 'JOBS', datetime.datetime(2022, 1, 12, 22, 41, 8, 30650)),
 ('XQK06680', 'DEPARTMENTS', datetime.datetime(2022, 1, 12, 22, 41, 8, 254844)),
 ('XQK06680', 'LOCATIONS', datetime.datetime(2022, 1, 12, 22, 41, 8, 488006)),
 ('XQK06680', 'PETSALE', datetime.datetime(2022, 1, 12, 20, 49, 37, 421802)),
 ('XQK06680', 'PETRESCUE', datetime.datetime(2022, 1, 15, 3, 47, 24, 127895)),
 ('XQK06680', 'INSTRUCTOR', datetime.datetime(2022, 1, 16, 23, 46, 2, 4331)),
 ('XQK06680', 'INTERNATIONAL_STUDENT_TEST_SCORES', datetime.datetime(2022, 1, 18, 8, 46, 24, 787113)),
 ('XQK06680', 'CHICAGO_SOCIOECONOMIC_DATA', datetime.datetime(2022, 1, 18, 14, 26, 31, 462088)),
 ('XQK06680', 'BOOKSHOP', datetime.datetime(2022, 1, 20, 7, 40, 24, 618304)),
 ('XQK06680', 'PUBLIC_SCHOOLS', datetime.datetime(2022, 1, 21, 2, 15, 36, 623115))]

In [5]:
%sql select * from SYSCAT.TABLES where TABNAME = 'PUBLIC_SCHOOLS'

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('XQK06680', 'PUBLIC_SCHOOLS', 'XQK06680', 'U', 'T', 'N', None, None, None, None, datetime.datetime(2022, 1, 21, 2, 15, 36, 623115), datetime.datetime(2022, 1, 21, 3, 32, 1, 383808), datetime.datetime(2022, 1, 21, 3, 32, 1, 383821), None, 78, 14, 1031, -1, -1, -1, -1, -1, -1, -1, -1, 'XQK06680SPACE1', None, None, 0, 0, 0, 0, 0, 0, 0, 'N', 'YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY', 1, ' ', '0', -1, 'N', ' ', None, 'R', ' ', 'N', '                                ', None, 'N', ' ', 'F', None, -1, 'N', 999, -1, -1.0, -1, -1.0, None, 1208, 'SYSIBM', 'IDENTITY', 'SYSIBM', 'IDENTITY', ' ', -1, datetime.datetime(2022, 1, 21, 2, 15, 36, 623115), 0, ' ', None, None, 'N', 'XQK06680', ' ', ' ', ' ', datetime.date(2022, 1, 21), ' ', 'N', 'R', 'N', -1.0, None)]

Double-click **here** for a hint

<!--
In Db2 the system catalog table called SYSCAT.TABLES contains the table metadata
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES where TABSCHEMA='YOUR-DB2-USERNAME'

or, you can retrieve list of all tables where the schema name is not one of the system created ones:

%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES \
      where TABSCHEMA not in ('SYSIBM', 'SYSCAT', 'SYSSTAT', 'SYSIBMADM', 'SYSTOOLS', 'SYSPUBLIC')
      
or, just query for a specifc table that you want to verify exists in the database
%sql select * from SYSCAT.TABLES where TABNAME = 'SCHOOLS'

-->


### Query the database system catalog to retrieve column metadata

##### The SCHOOLS table contains a large number of columns. How many columns does this table have?


In [6]:
# type in your query to retrieve the number of columns in the SCHOOLS table
%sql select count(*) from syscat.columns where tabname ='PUBLIC_SCHOOLS';

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(78,)]

Double-click **here** for a hint

<!--
In Db2 the system catalog table called SYSCAT.COLUMNS contains the column metadata
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from SYSCAT.COLUMNS where TABNAME = 'SCHOOLS'

-->


Now retrieve the the list of columns in SCHOOLS table and their column type (datatype) and length.


In [7]:
# type in your query to retrieve all column names in the SCHOOLS table along with their datatypes and length
# %sql select * from SYSCAT.COLUMNS where tabname='PUBLIC_SCHOOLS' order by colno;
%sql select colname,typename,length,colno from SYSCAT.COLUMNS where tabname='PUBLIC_SCHOOLS' order by colno;

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('SCHOOL_ID', 'INTEGER', 4, 0),
 ('NAME_OF_SCHOOL', 'VARCHAR', 64, 1),
 ('SCHOOL_TYPE', 'VARCHAR', 2, 2),
 ('STREET_ADDRESS', 'VARCHAR', 29, 3),
 ('CITY', 'VARCHAR', 7, 4),
 ('STATE', 'VARCHAR', 2, 5),
 ('ZIP_CODE', 'INTEGER', 4, 6),
 ('PHONE_NUMBER', 'VARCHAR', 14, 7),
 ('LINK', 'VARCHAR', 78, 8),
 ('NETWORK_MANAGER', 'VARCHAR', 40, 9),
 ('COLLABORATIVE_NAME', 'VARCHAR', 34, 10),
 ('ADEQUATE_YEARLY_PROGRESS_MADE_', 'VARCHAR', 3, 11),
 ('TRACK_SCHEDULE', 'VARCHAR', 12, 12),
 ('CPS_PERFORMANCE_POLICY_STATUS', 'VARCHAR', 16, 13),
 ('CPS_PERFORMANCE_POLICY_LEVEL', 'VARCHAR', 15, 14),
 ('HEALTHY_SCHOOL_CERTIFIED', 'VARCHAR', 3, 15),
 ('SAFETY_ICON', 'VARCHAR', 11, 16),
 ('SAFETY_SCORE', 'SMALLINT', 2, 17),
 ('FAMILY_INVOLVEMENT_ICON', 'VARCHAR', 11, 18),
 ('FAMILY_INVOLVEMENT_SCORE', 'VARCHAR', 3, 19),
 ('ENVIRONMENT_ICON', 'VARCHAR', 11, 20),
 ('ENVIRONMENT_SCORE', 'SMALLINT', 2, 21),
 ('INSTRUCTION_ICON', 'VARCHAR', 11, 22),
 ('INSTRUCTION_SCORE', 'SMALLINT', 2, 23),
 ('LEADERS_ICON', 'VARCHAR', 4, 24),
 ('LEADERS_SCORE', 'VARCHAR', 3, 25),
 ('TEACHERS_ICON', 'VARCHAR', 11, 26),
 ('TEACHERS_SCORE', 'VARCHAR', 3, 27),
 ('PARENT_ENGAGEMENT_ICON', 'VARCHAR', 7, 28),
 ('PARENT_ENGAGEMENT_SCORE', 'VARCHAR', 3, 29),
 ('PARENT_ENVIRONMENT_ICON', 'VARCHAR', 7, 30),
 ('PARENT_ENVIRONMENT_SCORE', 'VARCHAR', 3, 31),
 ('AVERAGE_STUDENT_ATTENDANCE', 'VARCHAR', 6, 32),
 ('RATE_OF_MISCONDUCTS__PER_100_STUDENTS_', 'DECIMAL', 5, 33),
 ('AVERAGE_TEACHER_ATTENDANCE', 'VARCHAR', 6, 34),
 ('INDIVIDUALIZED_EDUCATION_PROGRAM_COMPLIANCE_RATE', 'VARCHAR', 7, 35),
 ('PK_2_LITERACY__', 'VARCHAR', 4, 36),
 ('PK_2_MATH__', 'VARCHAR', 4, 37),
 ('GR3_5_GRADE_LEVEL_MATH__', 'VARCHAR', 4, 38),
 ('GR3_5_GRADE_LEVEL_READ__', 'VARCHAR', 4, 39),
 ('GR3_5_KEEP_PACE_READ__', 'VARCHAR', 4, 40),
 ('GR3_5_KEEP_PACE_MATH__', 'VARCHAR', 4, 41),
 ('GR6_8_GRADE_LEVEL_MATH__', 'VARCHAR', 4, 42),
 ('GR6_8_GRADE_LEVEL_READ__', 'VARCHAR', 4, 43),
 ('GR6_8_KEEP_PACE_MATH_', 'VARCHAR', 4, 44),
 ('GR6_8_KEEP_PACE_READ__', 'VARCHAR', 4, 45),
 ('GR_8_EXPLORE_MATH__', 'VARCHAR', 4, 46),
 ('GR_8_EXPLORE_READ__', 'VARCHAR', 4, 47),
 ('ISAT_EXCEEDING_MATH__', 'DECIMAL', 4, 48),
 ('ISAT_EXCEEDING_READING__', 'DECIMAL', 4, 49),
 ('ISAT_VALUE_ADD_MATH', 'DECIMAL', 3, 50),
 ('ISAT_VALUE_ADD_READ', 'DECIMAL', 3, 51),
 ('ISAT_VALUE_ADD_COLOR_MATH', 'VARCHAR', 6, 52),
 ('ISAT_VALUE_ADD_COLOR_READ', 'VARCHAR', 6, 53),
 ('STUDENTS_TAKING__ALGEBRA__', 'VARCHAR', 4, 54),
 ('STUDENTS_PASSING__ALGEBRA__', 'VARCHAR', 4, 55),
 ('9th Grade EXPLORE (2009)', 'VARCHAR', 4, 56),
 ('9th Grade EXPLORE (2010)', 'VARCHAR', 4, 57),
 ('10th Grade PLAN (2009)', 'VARCHAR', 4, 58),
 ('10th Grade PLAN (2010)', 'VARCHAR', 4, 59),
 ('NET_CHANGE_EXPLORE_AND_PLAN', 'VARCHAR', 3, 60),
 ('11th Grade Average ACT (2011)', 'VARCHAR', 4, 61),
 ('NET_CHANGE_PLAN_AND_ACT', 'VARCHAR', 3, 62),
 ('COLLEGE_ELIGIBILITY__', 'VARCHAR', 4, 63),
 ('GRADUATION_RATE__', 'VARCHAR', 4, 64),
 ('COLLEGE_ENROLLMENT_RATE__', 'VARCHAR', 4, 65),
 ('COLLEGE_ENROLLMENT', 'SMALLINT', 2, 66),
 ('GENERAL_SERVICES_ROUTE', 'SMALLINT', 2, 67),
 ('FRESHMAN_ON_TRACK_RATE__', 'VARCHAR', 4, 68),
 ('X_COORDINATE', 'DECIMAL', 13, 69),
 ('Y_COORDINATE', 'DECIMAL', 13, 70),
 ('LATITUDE', 'DECIMAL', 18, 71),
 ('LONGITUDE', 'DECIMAL', 18, 72),
 ('COMMUNITY_AREA_NUMBER', 'SMALLINT', 2, 73),
 ('COMMUNITY_AREA_NAME', 'VARCHAR', 22, 74),
 ('WARD', 'SMALLINT', 2, 75),
 ('POLICE_DISTRICT', 'SMALLINT', 2, 76),
 ('LOCATION', 'VARCHAR', 27, 77)]

Double-click **here** for the solution.

<!-- Solution:

%sql select COLNAME, TYPENAME, LENGTH from SYSCAT.COLUMNS where TABNAME = 'SCHOOLS'

or

%sql select distinct(NAME), COLTYPE, LENGTH from SYSIBM.SYSCOLUMNS where TBNAME = 'SCHOOLS'

-->


### Questions

1.  Is the column name for the "SCHOOL ID" attribute in upper or mixed case?
2.  What is the name of "Community Area Name" column in your table? Does it have spaces?
3.  Are there any columns in whose names the spaces and paranthesis (round brackets) have been replaced by the underscore character "\_"?


In [8]:
Ans 1-> 'School ID column is in Upper case'
Ans 2-> 'Name of "Community Area Name" is COMMUNITY_AREA_NUMBER, it has no spaces'
Ans 3-> 'No the space and paranthesis charachter are still present in the column namess'

SyntaxError: invalid syntax (<ipython-input-8-9ed5654fec31>, line 1)

## Problems

### Problem 1

##### How many Elementary Schools are in the dataset?


In [9]:
%sql select count(*) from PUBLIC_SCHOOLS group by 'Elementary, Middle, or High School'

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(566,)]

In [10]:
%sql select * from PUBLIC_SCHOOLS limit 4;

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(610038, 'Abraham Lincoln Elementary School', 'ES', '615 W Kemper Pl', 'Chicago', 'IL', 60614, '(773) 534-5720', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf', 'Fullerton Elementary Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 1', 'Yes', 'Very Strong', 99, 'Very Strong', '99', 'Strong', 74, 'Strong', 66, 'Weak', '65', 'Strong', '70', 'Strong', '56', 'Average', '47', '96.00%', Decimal('2.0'), '96.40%', '95.80%', '80.1', '43.3', '89.6', '84.9', '60.7', '62.6', '81.9', '85.2', '52', '62.4', '66.3', '77.9', Decimal('69.7'), Decimal('64.4'), Decimal('0.2'), Decimal('0.9'), 'Yellow', 'Green', '67.1', '54.5', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 813, 33, 'NDA', Decimal('1171699.458'), Decimal('1915829.428'), Decimal('41.92449696'), Decimal('-87.64452163'), 7, 'LINCOLN PARK', 43, 18, '(41.92449696, -87.64452163)'),
 (610281, 'Adam Clayton Powell Paideia Community Academy Elementary School', 'ES', '7511 S South Shore Dr', 'Chicago', 'IL', 60649, '(773) 535-6650', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610281.pdf', 'Skyway Elementary Network', 'SOUTH SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 54, 'Strong', '66', 'Strong', 74, 'Very Strong', 84, 'Weak', '63', 'Strong', '76', 'Weak', '46', 'Average', '50', '95.60%', Decimal('15.7'), '95.30%', '100.00%', '62.4', '51.7', '21.9', '15.1', '29', '42.8', '38.5', '27.4', '44.8', '42.7', '14.1', '34.4', Decimal('16.8'), Decimal('16.5'), Decimal('0.7'), Decimal('1.4'), 'Green', 'Green', '17.2', '27.3', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 521, 46, 'NDA', Decimal('1196129.985'), Decimal('1856209.466'), Decimal('41.76032435'), Decimal('-87.55673627'), 43, 'SOUTH SHORE', 7, 4, '(41.76032435, -87.55673627)'),
 (610185, 'Adlai E Stevenson Elementary School', 'ES', '8010 S Kostner Ave', 'Chicago', 'IL', 60652, '(773) 535-2280', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610185.pdf', 'Midway Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 2', 'No', 'Strong', 61, 'NDA', 'NDA', 'Average', 50, 'Weak', 36, 'Weak', 'NDA', 'NDA', 'NDA', 'Average', '47', 'Weak', '41', '95.70%', Decimal('2.3'), '94.70%', '98.30%', '53.7', '26.6', '38.3', '34.7', '43.7', '57.3', '48.8', '39.2', '46.8', '44', '7.5', '21.9', Decimal('18.3'), Decimal('15.5'), Decimal('-0.9'), Decimal('-1.0'), 'Red', 'Red', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 1324, 44, 'NDA', Decimal('1148427.165'), Decimal('1851012.215'), Decimal('41.74711093'), Decimal('-87.73170248'), 70, 'ASHBURN', 13, 8, '(41.74711093, -87.73170248)'),
 (609993, 'Agustin Lara Elementary Academy', 'ES', '4619 S Wolcott Ave', 'Chicago', 'IL', 60609, '(773) 535-4389', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_609993.pdf', 'Pershing Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 56, 'Average', '44', 'Average', 45, 'Weak', 37, 'Weak', '65', 'Average', '48', 'Average', '53', 'Strong', '58', '95.50%', Decimal('10.4'), '95.80%', '100.00%', '76.9', 'NDA', '26', '24.7', '61.8', '49.7', '39.2', '27.2', '69.7', '60.6', '9.1', '18.2', Decimal('11.1'), Decimal('9.6'), Decimal('0.9'), Decimal('2.4'), 'Green', 'Green', '42.9', '25', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 556, 42, 'NDA', Decimal('1164504.290'), Decimal('1873959.199'), Decimal('41.80975690'), Decimal('-87.67214460'), 61, 'NEW CITY', 20, 9, '(41.8097569, -87.6721446)')]

In [11]:
#%sql ALTER TABLE PUBLIC_SCHOOLS RENAME COLUMN \"Elementary, Middle, or High School\" to SCHOOL_TYPE;
%sql select * from PUBLIC_SCHOOLS limit 4

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(610038, 'Abraham Lincoln Elementary School', 'ES', '615 W Kemper Pl', 'Chicago', 'IL', 60614, '(773) 534-5720', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf', 'Fullerton Elementary Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 1', 'Yes', 'Very Strong', 99, 'Very Strong', '99', 'Strong', 74, 'Strong', 66, 'Weak', '65', 'Strong', '70', 'Strong', '56', 'Average', '47', '96.00%', Decimal('2.0'), '96.40%', '95.80%', '80.1', '43.3', '89.6', '84.9', '60.7', '62.6', '81.9', '85.2', '52', '62.4', '66.3', '77.9', Decimal('69.7'), Decimal('64.4'), Decimal('0.2'), Decimal('0.9'), 'Yellow', 'Green', '67.1', '54.5', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 813, 33, 'NDA', Decimal('1171699.458'), Decimal('1915829.428'), Decimal('41.92449696'), Decimal('-87.64452163'), 7, 'LINCOLN PARK', 43, 18, '(41.92449696, -87.64452163)'),
 (610281, 'Adam Clayton Powell Paideia Community Academy Elementary School', 'ES', '7511 S South Shore Dr', 'Chicago', 'IL', 60649, '(773) 535-6650', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610281.pdf', 'Skyway Elementary Network', 'SOUTH SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 54, 'Strong', '66', 'Strong', 74, 'Very Strong', 84, 'Weak', '63', 'Strong', '76', 'Weak', '46', 'Average', '50', '95.60%', Decimal('15.7'), '95.30%', '100.00%', '62.4', '51.7', '21.9', '15.1', '29', '42.8', '38.5', '27.4', '44.8', '42.7', '14.1', '34.4', Decimal('16.8'), Decimal('16.5'), Decimal('0.7'), Decimal('1.4'), 'Green', 'Green', '17.2', '27.3', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 521, 46, 'NDA', Decimal('1196129.985'), Decimal('1856209.466'), Decimal('41.76032435'), Decimal('-87.55673627'), 43, 'SOUTH SHORE', 7, 4, '(41.76032435, -87.55673627)'),
 (610185, 'Adlai E Stevenson Elementary School', 'ES', '8010 S Kostner Ave', 'Chicago', 'IL', 60652, '(773) 535-2280', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610185.pdf', 'Midway Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 2', 'No', 'Strong', 61, 'NDA', 'NDA', 'Average', 50, 'Weak', 36, 'Weak', 'NDA', 'NDA', 'NDA', 'Average', '47', 'Weak', '41', '95.70%', Decimal('2.3'), '94.70%', '98.30%', '53.7', '26.6', '38.3', '34.7', '43.7', '57.3', '48.8', '39.2', '46.8', '44', '7.5', '21.9', Decimal('18.3'), Decimal('15.5'), Decimal('-0.9'), Decimal('-1.0'), 'Red', 'Red', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 1324, 44, 'NDA', Decimal('1148427.165'), Decimal('1851012.215'), Decimal('41.74711093'), Decimal('-87.73170248'), 70, 'ASHBURN', 13, 8, '(41.74711093, -87.73170248)'),
 (609993, 'Agustin Lara Elementary Academy', 'ES', '4619 S Wolcott Ave', 'Chicago', 'IL', 60609, '(773) 535-4389', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_609993.pdf', 'Pershing Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 56, 'Average', '44', 'Average', 45, 'Weak', 37, 'Weak', '65', 'Average', '48', 'Average', '53', 'Strong', '58', '95.50%', Decimal('10.4'), '95.80%', '100.00%', '76.9', 'NDA', '26', '24.7', '61.8', '49.7', '39.2', '27.2', '69.7', '60.6', '9.1', '18.2', Decimal('11.1'), Decimal('9.6'), Decimal('0.9'), Decimal('2.4'), 'Green', 'Green', '42.9', '25', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 556, 42, 'NDA', Decimal('1164504.290'), Decimal('1873959.199'), Decimal('41.80975690'), Decimal('-87.67214460'), 61, 'NEW CITY', 20, 9, '(41.8097569, -87.6721446)')]

In [12]:
%sql select SCHOOL_TYPE,count(*) as count_of_school_type from PUBLIC_SCHOOLS GROUP BY SCHOOL_TYPE

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('ES', 462), ('HS', 93), ('MS', 11)]

Double-click **here** for a hint

<!--
Which column specifies the school type e.g. 'ES', 'MS', 'HS'? ("Elementary School, Middle School, High School")
-->


Double-click **here** for another hint

<!--
Does the column name have mixed case, spaces or other special characters?
If so, ensure you use double quotes around the "Name of the Column"
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from SCHOOLS where "Elementary, Middle, or High School" = 'ES'

Correct answer: 462

-->


### Problem 2

##### What is the highest Safety Score?


In [13]:
%sql select max(safety_score) from PUBLIC_SCHOOLS

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(99,)]

Double-click **here** for a hint

<!--
Use the MAX() function
-->


Double-click **here** for the solution.

<!-- Hint:

%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from SCHOOLS

Correct answer: 99
-->


### Problem 3

##### Which schools have highest Safety Score?


In [14]:
%sql select name_of_school,safety_score from PUBLIC_SCHOOLS where safety_score=(select max(safety_score) from PUBLIC_SCHOOLS)

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('Abraham Lincoln Elementary School', 99),
 ('Alexander Graham Bell Elementary School', 99),
 ('Annie Keller Elementary Gifted Magnet School', 99),
 ('Augustus H Burley Elementary School', 99),
 ('Edgar Allan Poe Elementary Classical School', 99),
 ('Edgebrook Elementary School', 99),
 ('Ellen Mitchell Elementary School', 99),
 ('James E McDade Elementary Classical School', 99),
 ('James G Blaine Elementary School', 99),
 ('LaSalle Elementary Language Academy', 99),
 ('Mary E Courtenay Elementary Language Arts Center', 99),
 ('Northside College Preparatory High School', 99),
 ('Northside Learning Center High School', 99),
 ('Norwood Park Elementary School', 99),
 ('Oriole Park Elementary School', 99),
 ('Sauganash Elementary School', 99),
 ('Stephen Decatur Classical Elementary School', 99),
 ('Talman Elementary School', 99),
 ('Wildwood Elementary School', 99)]

Double-click **here** for the solution.

<!-- Solution:
In the previous problem we found out that the highest Safety Score is 99, so we can use that as an input in the where clause:

%sql select Name_of_School, Safety_Score from SCHOOLS where Safety_Score = 99

or, a better way:

%sql select Name_of_School, Safety_Score from SCHOOLS where \
  Safety_Score= (select MAX(Safety_Score) from SCHOOLS)


Correct answer: several schools with with Safety Score of 99.
-->


### Problem 4

##### What are the top 10 schools with the highest "Average Student Attendance"?


In [15]:
#%sql select name_of_school,average_student_attendance from PUBLIC_SCHOOLS order by average_student_attendance desc limit 10
%sql select name_of_school,average_student_attendance from PUBLIC_SCHOOLS order by average_student_attendance desc nulls last limit 10

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('John Charles Haines Elementary School', '98.40%'),
 ('James Ward Elementary School', '97.80%'),
 ('Edgar Allan Poe Elementary Classical School', '97.60%'),
 ('Orozco Fine Arts & Sciences Elementary School', '97.60%'),
 ('Rachel Carson Elementary School', '97.60%'),
 ('Annie Keller Elementary Gifted Magnet School', '97.50%'),
 ('Andrew Jackson Elementary Language Academy', '97.40%'),
 ('Lenart Elementary Regional Gifted Center', '97.40%'),
 ('Disney II Magnet School', '97.30%'),
 ('John H Vanderpoel Elementary Magnet School', '97.20%')]

Double-click **here** for the solution.

<!-- Solution:

%sql select Name_of_School, Average_Student_Attendance from SCHOOLS \
    order by Average_Student_Attendance desc nulls last limit 10 

-->


### Problem 5

##### Retrieve the list of 5 Schools with the lowest Average Student Attendance sorted in ascending order based on attendance


In [16]:
%sql select name_of_school, average_student_attendance \
from PUBLIC_SCHOOLS \
order by average_student_attendance nulls last limit 5

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Dyett High School', '62.50%'),
 ('Wendell Phillips Academy High School', '63.00%'),
 ('Orr Academy High School', '66.30%')]

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     order by Average_Student_Attendance \
     fetch first 5 rows only

-->


### Problem 6

##### Now remove the '%' sign from the above result set for Average Student Attendance column


In [29]:
%sql select name_of_school, REPLACE(average_student_attendance,'%','') \
from PUBLIC_SCHOOLS \
order by average_student_attendance fetch first 5 rows only

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('Richard T Crane Technical Preparatory High School', '57.90'),
 ('Barbara Vick Early Childhood & Family Center', '60.90'),
 ('Dyett High School', '62.50'),
 ('Wendell Phillips Academy High School', '63.00'),
 ('Orr Academy High School', '66.30')]

Double-click **here** for a hint

<!--
Use the REPLACE() function to replace '%' with ''
See documentation for this function at:
https://www.ibm.com/support/knowledgecenter/en/SSEPGG_10.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000843.html
-->


Double-click **here** for the solution.

<!-- Hint:

%sql SELECT Name_of_School, REPLACE(Average_Student_Attendance, '%', '') \
     from SCHOOLS \
     order by Average_Student_Attendance \
     fetch first 5 rows only

-->


### Problem 7

##### Which Schools have Average Student Attendance lower than 70%?


In [31]:
%sql select name_of_school,average_student_attendance from public_schools where average_student_attendance<'70%'

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Chicago Vocational Career Academy High School', '68.80%'),
 ('Dyett High School', '62.50%'),
 ('Manley Career Academy High School', '66.80%'),
 ('Orr Academy High School', '66.30%'),
 ('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Roberto Clemente Community Academy High School', '69.60%'),
 ('Wendell Phillips Academy High School', '63.00%')]

Double-click **here** for a hint

<!--
The datatype of the "Average_Student_Attendance" column is varchar.
So you cannot use it as is in the where clause for a numeric comparison.
First use the CAST() function to cast it as a DECIMAL or DOUBLE
e.g. CAST("Column_Name" as DOUBLE)
or simply: DECIMAL("Column_Name")
-->


Double-click **here** for another hint

<!--
Don't forget the '%' age sign needs to be removed before casting
-->


Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     where CAST ( REPLACE(Average_Student_Attendance, '%', '') AS DOUBLE ) < 70 \
     order by Average_Student_Attendance
     
or,

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     where DECIMAL ( REPLACE(Average_Student_Attendance, '%', '') ) < 70 \
     order by Average_Student_Attendance

-->


### Problem 8

##### Get the total College Enrollment for each Community Area


In [32]:
%sql select community_area_name,sum(college_enrollment) from public_schools group by community_area_name

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('ALBANY PARK', 6864),
 ('ARCHER HEIGHTS', 4823),
 ('ARMOUR SQUARE', 1458),
 ('ASHBURN', 6483),
 ('AUBURN GRESHAM', 4175),
 ('AUSTIN', 10933),
 ('AVALON PARK', 1522),
 ('AVONDALE', 3640),
 ('BELMONT CRAGIN', 14386),
 ('BEVERLY', 1636),
 ('BRIDGEPORT', 3167),
 ('BRIGHTON PARK', 9647),
 ('BURNSIDE', 549),
 ('CALUMET HEIGHTS', 1568),
 ('CHATHAM', 5042),
 ('CHICAGO LAWN', 7086),
 ('CLEARING', 2085),
 ('DOUGLAS', 4670),
 ('DUNNING', 4568),
 ('EAST GARFIELD PARK', 5337),
 ('EAST SIDE', 5305),
 ('EDGEWATER', 4600),
 ('EDISON PARK', 910),
 ('ENGLEWOOD', 6832),
 ('FOREST GLEN', 1431),
 ('FULLER PARK', 531),
 ('GAGE PARK', 9915),
 ('GARFIELD RIDGE', 4552),
 ('GRAND BOULEVARD', 2809),
 ('GREATER GRAND CROSSING', 4051),
 ('HEGEWISCH', 963),
 ('HERMOSA', 3975),
 ('HUMBOLDT PARK', 8620),
 ('HYDE PARK', 1930),
 ('IRVING PARK', 7764),
 ('JEFFERSON PARK', 1755),
 ('KENWOOD', 4287),
 ('LAKE VIEW', 7055),
 ('LINCOLN PARK', 5615),
 ('LINCOLN SQUARE', 4132),
 ('LOGAN SQUARE', 7351),
 ('LOOP', 871),
 ('LOWER WEST SIDE', 7257),
 ('MCKINLEY PARK', 1552),
 ('MONTCLARE', 1317),
 ('MORGAN PARK', 3271),
 ('MOUNT GREENWOOD', 2091),
 ('NEAR NORTH SIDE', 3362),
 ('NEAR SOUTH SIDE', 1378),
 ('NEAR WEST SIDE', 7975),
 ('NEW CITY', 7922),
 ('NORTH CENTER', 7541),
 ('NORTH LAWNDALE', 5146),
 ('NORTH PARK', 4210),
 ('NORWOOD PARK', 6469),
 ('OAKLAND', 140),
 ('OHARE', 786),
 ('PORTAGE PARK', 6954),
 ('PULLMAN', 1620),
 ('RIVERDALE', 1547),
 ('ROGERS PARK', 4068),
 ('ROSELAND', 7020),
 ('SOUTH CHICAGO', 4043),
 ('SOUTH DEERING', 1859),
 ('SOUTH LAWNDALE', 14793),
 ('SOUTH SHORE', 4543),
 ('UPTOWN', 4388),
 ('WASHINGTON HEIGHTS', 4006),
 ('WASHINGTON PARK', 2648),
 ('WEST ELSDON', 3700),
 ('WEST ENGLEWOOD', 5946),
 ('WEST GARFIELD PARK', 2622),
 ('WEST LAWN', 4207),
 ('WEST PULLMAN', 3240),
 ('WEST RIDGE', 8197),
 ('WEST TOWN', 9429),
 ('WOODLAWN', 4206)]

Double-click **here** for a hint

<!--
Verify the exact name of the Enrollment column in the database
Use the SUM() function to add up the Enrollments for each Community Area
-->


Double-click **here** for another hint

<!--
Don't forget to group by the Community Area
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from SCHOOLS \
   group by Community_Area_Name 

-->


### Problem 9

##### Get the 5 Community Areas with the least total College Enrollment  sorted in ascending order


In [33]:
%sql select community_area_name,sum(college_enrollment) as total_college_enrollment from public_schools group by community_area_name order by total_college_enrollment fetch first 5 rows only

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('OAKLAND', 140),
 ('FULLER PARK', 531),
 ('BURNSIDE', 549),
 ('OHARE', 786),
 ('LOOP', 871)]

Double-click **here** for a hint

<!--
Order the previous query and limit the number of rows you fetch
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from SCHOOLS \
   group by Community_Area_Name \
   order by TOTAL_ENROLLMENT asc \
   fetch first 5 rows only

-->


### Problem 10

##### List 5 schools with lowest safety score.


In [39]:
%sql select name_of_school,safety_score from public_schools order by safety_score fetch first 5 rows only
# limit can also be used

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('Edmond Burke Elementary School', 1),
 ("Luke O'Toole Elementary School", 5),
 ('George W Tilton Elementary School', 6),
 ('Foster Park Elementary School', 11),
 ('Emil G Hirsch Metropolitan High School', 13)]

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT name_of_school, safety_score \
FROM schools \
ORDER BY safety_score \
LIMIT 5
-->


### Problem 11

##### Get the hardship index for the community area which has College Enrollment of 4368


In [50]:
%sql select hardship_index,ps.community_area_name from public_schools ps, chicago_socioeconomic_data se where ps.community_area_number=se.ca and college_enrollment=4368

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(6.0, 'NORTH CENTER')]

Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%%sql 
select hardship_index 
   from chicago_socioeconomic_data CD, schools CPS 
   where CD.ca = CPS.community_area_number 
      and college_enrollment = 4368

-->


### Problem 12

##### Get the hardship index for the community area which has the school with the  highest enrollment.


In [53]:
%sql select hardship_index,PS.community_area_name from public_schools PS,chicago_socioeconomic_data CD where PS.community_area_number=CD.ca and college_enrollment = (select max(college_enrollment) from public_schools)

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(6.0, 'NORTH CENTER')]

In [57]:
%sql select school_id from public_schools

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[(610038,),
 (610281,),
 (610185,),
 (609993,),
 (610513,),
 (610212,),
 (609720,),
 (610342,),
 (610524,),
 (610209,),
 (609799,),
 (609947,),
 (609963,),
 (610210,),
 (609808,),
 (610028,),
 (610098,),
 (609788,),
 (610334,),
 (610131,),
 (610316,),
 (609780,),
 (610339,),
 (610320,),
 (609837,),
 (610060,),
 (609869,),
 (610084,),
 (610229,),
 (610277,),
 (609951,),
 (610161,),
 (610171,),
 (610037,),
 (609887,),
 (609836,),
 (610268,),
 (609941,),
 (610287,),
 (609820,),
 (400018,),
 (610169,),
 (610501,),
 (609786,),
 (609871,),
 (609922,),
 (609764,),
 (610265,),
 (610290,),
 (610076,),
 (610150,),
 (610175,),
 (610087,),
 (609969,),
 (610106,),
 (610317,),
 (610381,),
 (609821,),
 (609827,),
 (610353,),
 (609729,),
 (610039,),
 (610238,),
 (609839,),
 (610345,),
 (610148,),
 (609679,),
 (610005,),
 (609879,),
 (609966,),
 (610366,),
 (610030,),
 (609996,),
 (609833,),
 (609732,),
 (609875,),
 (609986,),
 (609813,),
 (609883,),
 (610194,),
 (609897,),
 (610248,),
 (610340,),
 (609753,),
 (609754,),
 (609674,),
 (609899,),
 (609705,),
 (609863,),
 (609790,),
 (610347,),
 (610499,),
 (610170,),
 (610004,),
 (609781,),
 (609873,),
 (610063,),
 (609804,),
 (610083,),
 (610380,),
 (609870,),
 (609835,),
 (610223,),
 (610221,),
 (609704,),
 (610402,),
 (609885,),
 (609859,),
 (610515,),
 (610077,),
 (610198,),
 (610533,),
 (609751,),
 (610266,),
 (610188,),
 (610352,),
 (610254,),
 (609736,),
 (610132,),
 (609901,),
 (610523,),
 (609819,),
 (609815,),
 (610246,),
 (609881,),
 (609862,),
 (609909,),
 (610012,),
 (609904,),
 (610006,),
 (609735,),
 (610315,),
 (609708,),
 (609913,),
 (610233,),
 (610227,),
 (610195,),
 (610119,),
 (609852,),
 (610235,),
 (609682,),
 (610073,),
 (609712,),
 (609973,),
 (610065,),
 (609916,),
 (610203,),
 (610543,),
 (610137,),
 (609908,),
 (609918,),
 (609874,),
 (610319,),
 (610057,),
 (610541,),
 (610308,),
 (610120,),
 (609917,),
 (610258,),
 (610096,),
 (609924,),
 (609927,),
 (610215,),
 (610067,),
 (610114,),
 (610020,),
 (610112,),
 (609800,),
 (609939,),
 (609926,),
 (609958,),
 (610144,),
 (610165,),
 (610503,),
 (609854,),
 (610245,),
 (609930,),
 (610163,),
 (610010,),
 (609737,),
 (609709,),
 (610009,),
 (610072,),
 (610293,),
 (609779,),
 (610062,),
 (610196,),
 (609849,),
 (609761,),
 (610305,),
 (610048,),
 (610139,),
 (609795,),
 (609760,),
 (609845,),
 (610219,),
 (609739,),
 (609900,),
 (609693,),
 (610202,),
 (609920,),
 (610187,),
 (610285,),
 (610383,),
 (609857,),
 (610241,),
 (609741,),
 (609726,),
 (610182,),
 (610068,),
 (610124,),
 (610192,),
 (610158,),
 (609974,),
 (609985,),
 (610122,),
 (609972,),
 (609856,),
 (610040,),
 (610092,),
 (610279,),
 (609861,),
 (610297,),
 (609798,),
 (609768,),
 (609850,),
 (610052,),
 (610058,),
 (609713,),
 (610390,),
 (610110,),
 (610273,),
 (610384,),
 (610078,),
 (609848,),
 (610532,),
 (610125,),
 (610276,),
 (610103,),
 (610256,),
 (609797,),
 (609766,),
 (609912,),
 (610070,),
 (610066,),
 (609803,),
 (609983,),
 (610047,),
 (610074,),
 (610200,),
 (610107,),
 (609891,),
 (610044,),
 (610174,),
 (610213,),
 (610217,),
 (610274,),
 (609772,),
 (610093,),
 (609796,),
 (610031,),
 (610271,),
 (609932,),
 (610173,),
 (609942,),
 (610216,),
 (609789,),
 (609894,),
 (610089,),
 (610243,),
 (609829,),
 (609866,),
 (609893,),
 (609959,),
 (610176,),
 (609898,),
 (609919,),
 (609718,),
 (610263,),
 (610228,),
 (609694,),
 (609971,),
 (609975,),
 (609964,),
 (610026,),
 (610207,),
 (609782,),
 (610126,),
 (610054,),
 (609723,),
 (609710,),
 (609954,),
 (610180,),
 (610199,),
 (610111,),
 (609775,),
 (610184,),
 (610269,),
 (610130,),
 (609864,),
 (609937,),
 (610225,),
 (609828,),
 (610159,),
 (609865,),
 (610313,),
 (609950,),
 (610242,),
 (609938,),
 (610041,),
 (610015,),
 (609925,),
 (610043,),
 (610189,),
 (610218,),
 (610019,),
 (610129,),
 (610022,),
 (610253,),
 (610000,),
 (610280,),
 (610016,),
 (609716,),
 (609746,),
 (609719,),
 (610368,),
 (610033,),
 (610520,),
 (609910,),
 (610133,),
 (610034,),
 (610024,),
 (609907,),
 (61

In [56]:
%sql select distinct leaders_score,leaders_icon from public_schools

 * ibm_db_sa://xqk06680:***@0c77d6f2-5da9-48a9-81f8-86b520b87518.bs2io90l08kqb1od8lcg.databases.appdomain.cloud:31198/BLUDB;security=SSL;
Done.


[('10', 'Weak'),
 ('14', 'Weak'),
 ('15', 'Weak'),
 ('16', 'Weak'),
 ('17', 'Weak'),
 ('19', 'Weak'),
 ('20', 'Weak'),
 ('21', 'Weak'),
 ('22', 'Weak'),
 ('24', 'Weak'),
 ('25', 'Weak'),
 ('26', 'Weak'),
 ('27', 'Weak'),
 ('28', 'Weak'),
 ('29', 'Weak'),
 ('30', 'Weak'),
 ('31', 'Weak'),
 ('32', 'Weak'),
 ('33', 'Weak'),
 ('34', 'Weak'),
 ('35', 'Weak'),
 ('36', 'Weak'),
 ('37', 'Weak'),
 ('38', 'Weak'),
 ('39', 'Weak'),
 ('4', 'Weak'),
 ('40', 'Weak'),
 ('41', 'Weak'),
 ('42', 'Weak'),
 ('43', 'Weak'),
 ('44', 'Weak'),
 ('45', 'Weak'),
 ('46', 'Weak'),
 ('47', 'Weak'),
 ('48', 'Weak'),
 ('49', 'Weak'),
 ('50', 'Weak'),
 ('51', 'Weak'),
 ('52', 'Weak'),
 ('53', 'Weak'),
 ('54', 'Weak'),
 ('55', 'Weak'),
 ('56', 'Weak'),
 ('57', 'Weak'),
 ('58', 'Weak'),
 ('59', 'Weak'),
 ('60', 'Weak'),
 ('61', 'Weak'),
 ('62', 'Weak'),
 ('63', 'Weak'),
 ('64', 'Weak'),
 ('65', 'Weak'),
 ('66', 'Weak'),
 ('67', 'Weak'),
 ('68', 'Weak'),
 ('69', 'Weak'),
 ('70', 'Weak'),
 ('71', 'Weak'),
 ('72', 'Weak'),
 ('73', 'Weak'),
 ('74', 'Weak'),
 ('75', 'Weak'),
 ('76', 'Weak'),
 ('77', 'Weak'),
 ('79', 'Weak'),
 ('80', 'Weak'),
 ('81', 'Weak'),
 ('82', 'Weak'),
 ('83', 'Weak'),
 ('84', 'Weak'),
 ('85', 'Weak'),
 ('86', 'Weak'),
 ('88', 'Weak'),
 ('90', 'Weak'),
 ('92', 'Weak'),
 ('94', 'Weak'),
 ('97', 'Weak'),
 ('99', 'Weak'),
 ('NDA', 'Weak')]

Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%sql select ca, community_area_name, hardship_index from chicago_socioeconomic_data \
   where ca in \
   ( select community_area_number from schools order by college_enrollment desc limit 1 )

-->


## Author

<a href="https://www.linkedin.com/in/ravahuja/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Rav Ahuja</a>

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                        |
| ----------------- | ------- | ----------------- | ----------------------------------------- |
| 2021-07-09        | 2.4     | Malika            | Updated connection string                 |
| 2021-05-19        | 2.3     | Lakshmi Holla     | Updated question                          |
| 2021-04-20        | 2.2     | Malika            | Added the libraries                       |
| 2020-11-27        | 2.1     | Sannareddy Ramesh | Modified data sets and added new problems |
| 2020-08-28        | 2.0     | Lavanya           | Moved lab to course repo in GitLab        |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
